### DISCUSSION TITLE
Moving Effect to a separate target?

### DISCUSSION THREAD

**Discussion by lukeredpath** [pointfreeco/swift-composable-architecture#1080] [Ideas]
Our app uses a highly modularised architecture similar to isowords - we have two packages, a features package and a "core" package.

To get the most out of modularisation it helps to try and keep your dependency graph under control, particularly avoiding large unfocused modules that pull in lots of other dependencies, otherwise you can end up in dependency hell.

A lot of our `XXXClient` type libraries are modelled around the TCA `Effect` type - this makes a lot of sense as we use TCA in our app and the type has some nice additional APIs for creating effects and it also avoids the need for creating ad-hoc effects or `eraseToEffect()` calls in our reducers. Unfortunately this means all of these libraries need to compile the entire TCA library just to use this single type. This isn't a big deal when building these libraries in the context of some bigger target like a feature module as we'd be compiling TCA anyway but when you just want to build your core library on its own, it seems less efficient than it could be.

In theory Xcode should be caching these dependencies but in practice the Xcode build system just doesn't work as well as it should in my experience. 

Given how common a approach this must be, would it make sense to create a new `Effects` target and library in the TCA package so that our client targets can depend on this alone?

**Comment by stephencelis** [comment]
@lukeredpath We've been recently working on some refinements to the effect story in TCA and one thing we discussed was extracting the `Effect` type for this very reason. After spending some time with the problem, though, we've come to an alternate solution that we think works better for some of the longer-term goals we have for TCA.

We hope to have all the details worked out in the coming weeks!

In the meantime, you might be able to achieve what you're looking for by changing your clients to return `AnyPublisher`, instead. The downside is losing some `Effect` helpers (though you could recover them on `AnyPublisher` with a little work), and added `eraseToEffect`s in your reducers, but it might serve as a good option while we prepare our experiments.

**Reply by lukeredpath** [reply]
Using `AnyPublisher` was something I was hoping to avoid - we can wait to see what you have in the pipeline. Happy to "beta test" anything you have planned if you want anything taken for a spin in a real significantly sized codebase.

**Reply by stephencelis** [reply]
We're still working out some details but will share a preview when it's ready!

**Reply by JacksonUtsch** [reply]
I have a set of dependencies that utilize `AnyPublisher` to avoid the common TCA dependency. Otherwise it could create some annoying version conflicts when trying to update, also can migrate to any combine based redux system if need be. In those I've implemented internal cancel actions and other functionality if need be. Works for me.

**Reply by lukeredpath** [reply]
@JacksonUtsch thanks - I know that I can use `AnyPublisher` but I don't want to do this as it would require me to erase to publishers and then erase back to effects again everywhere. Its not worth the code churn and migrating to another system isn't really a concern for me.

**Reply by JacksonUtsch** [reply]
Pardon the excessiveness.. But in addition I also do this to change the type signature of my reducers to utilize failure types which I find very nice in propagating global errors while keeping their identity. I hope TCA adopts this at some point. Then we don't have to handle errors in isolation.
